compiler/modules/front.syntax/tests/posTree.bal

```diff
@@ -0,0 +1,585 @@
+    childNodes.push({ name: defn.name, pos: defn.namePos });
+    childNodes.push(typeDescToNode(defn.td));
+    childNodes.push(SEMICOLON_NODE);
+    return { childNodes, astNode: defn };
+}
+
+function functionDefnToNode(FunctionDefn defn) returns SyntaxNode {
+    SyntaxNode[] childNodes = moduleLevelDefnPrefix(defn);
+    childNodes.push({ token: "function" });
+    childNodes.push({ name: defn.name, pos: defn.namePos });
+    childNodes.push(functionTypeDescToNode(defn.typeDesc));
+    childNodes.push(...stmtBlockToNodes(defn.body));
+    return { childNodes, astNode: defn };
+}
+
+function moduleLevelDefnPrefix(ModuleLevelDefn defn) returns SyntaxNode[] {
+    SyntaxNode[] childNodes = [];
+    if defn.vis == "public" {
+        childNodes.push({ token: "public", pos: defn.startPos });
+    }
+    return childNodes;
+}
+
+function functionTypeDescToNode(FunctionTypeDesc td) returns SyntaxNode {
+    SyntaxNode[] childNodes = [{ token: "(", pos: td.startPos }];
+    foreach int i in 0 ..< td.params.length() {
+        if i > 0 {
+            childNodes.push({ token: "," });
+        }
+        childNodes.push(functionTypeParamToNode(td.params[i]));
+    }
+    childNodes.push({ token: ")" });
+    TypeDesc retTd = td.ret;
+    // pr-todo: this will mess "returns null" functions
+    if !(retTd is BuiltinTypeDesc && retTd.builtinTypeName == "null") {
+        childNodes.push({ token: "returns" }, typeDescToNode(retTd));
+    }
+    return { childNodes, astNode: td };
+}
+
+function functionTypeParamToNode(FunctionTypeParam param) returns SyntaxNode {
+    SyntaxNode[] childNodes = [typeDescToNode(param.td)];
+    string? name = param.name;
+    if name != () {
+        childNodes.push({ name, pos:param.namePos });
+    }
+    return { childNodes, astNode: param };
+}
+
+function stmtBlockToNodes(StmtBlock block) returns SyntaxNode[] {
+    SyntaxNode[] childNodes = [{ token: "{", pos: block.startPos }];
+    foreach Stmt stmt in block.stmts {
+        childNodes.push(stmtToNode(stmt));
+    }
+    childNodes.push({ token: "}", pos: block.endPos });
+    return childNodes;
+}
+
+// function nodeFromStmt(Stmt stmt) returns SyntaxNode {
+//     if stmt is VarDeclStmt {
+//         return {
+//             astNode: stmt,
+//             childNodes: [ nodeFromTypeDesc(stmt.td), nodeFromExpr(stmt.initExpr) ]
+//         };
+//     }
+//     else if
+// }
+//
+// function nodesFromStmtBlock(StmtBlock block) returns SyntaxNode[] {
+//     SyntaxNode[] nodes = [];
+//     // Add terminal for `{`
+//     foreach var stmt in block.stmts {
+//         nodes.push(nodeFromStmt(stmt));
+//
+//     }
+//
+//     // Add terminal for `}` with closeBracePos
+//     return nodes;
+// }
+//
+// function nodeFromExpr(Expr expr) returns SyntaxNode {
+//
+// }
+//
+// function nodeFromTypeDesc(TypeDesc td) returns SyntaxNode {
+//
+// }
+
+function stmtToNode(Stmt stmt) returns NonTerminalSyntaxNode {
+    SyntaxNode[] childNodes;
+    if stmt is VarDeclStmt {
```

Comment:
Try to use a more functional style.

Have a helper function

```
nonTerminalSyntaxNode(AstNode ast, SyntaxNode... nodes) returns NonTerminalSyntaxNode
```

Use that rather than appending to childNodes. Each `if` will just do `return createNonTerminal(stmt, XXX);` `XXX` can use `...expr` as needed.

Maybe add another helper that also adds the semi-colon.

Consider having a separate function for each kind of stmt e.g. syntaxNodeFromVarDeclStmt.